google/resource_app_engine_application.go

```diff
@@ -0,0 +1,278 @@
+	}
+
+	d.SetId(project)
+
+	// Wait for the operation to complete
+	waitErr := appEngineOperationWait(config.clientAppEngine, op, project, "App Engine app to create")
+	if waitErr != nil {
+		return waitErr
+	}
+	log.Printf("[DEBUG] Created App Engine App")
+
+	return resourceAppEngineApplicationRead(d, meta)
+}
+
+func resourceAppEngineApplicationRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	pid := d.Id()
+
+	app, err := config.clientAppEngine.Apps.Get(pid).Do()
+	if err != nil && !isGoogleApiErrorWithCode(err, 404) {
+		return fmt.Errorf("Error retrieving App Engine application %q: %s", pid, err.Error())
+	} else if isGoogleApiErrorWithCode(err, 404) {
+		log.Printf("[WARN] App Engine application %q not found, removing from state", pid)
+		d.SetId("")
+		return nil
+	}
+	d.Set("auth_domain", app.AuthDomain)
+	d.Set("code_bucket", app.CodeBucket)
+	d.Set("default_bucket", app.DefaultBucket)
+	d.Set("default_hostname", app.DefaultHostname)
+	d.Set("location_id", app.LocationId)
+	d.Set("name", app.Name)
+	d.Set("serving_status", app.ServingStatus)
+	d.Set("project", pid)
+	dispatchRules, err := flattenAppEngineApplicationDispatchRules(app.DispatchRules)
+	if err != nil {
+		return err
+	}
+	err = d.Set("url_dispatch_rule", dispatchRules)
+	if err != nil {
+		return fmt.Errorf("Error setting dispatch rules in state. This is a bug, please report it at https://github.com/terraform-providers/terraform-provider-google/issues. Error is:\n%s", err.Error())
+	}
+	featureSettings, err := flattenAppEngineApplicationFeatureSettings(app.FeatureSettings)
+	if err != nil {
+		return err
+	}
+	err = d.Set("feature_settings", featureSettings)
+	if err != nil {
+		return fmt.Errorf("Error setting feature settings in state. This is a bug, please report it at https://github.com/terraform-providers/terraform-provider-google/issues. Error is:\n%s", err.Error())
+	}
+	return nil
+}
+
+func resourceAppEngineApplicationUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	pid := d.Id()
+	app, err := expandAppEngineApplication(d, pid)
+	if err != nil {
+		return err
+	}
+	log.Printf("[DEBUG] Updating App Engine App")
+	op, err := config.clientAppEngine.Apps.Patch(pid, app).UpdateMask("authDomain,servingStatus,featureSettings.splitHealthChecks").Do()
+	if err != nil {
+		return fmt.Errorf("Error updating App Engine application: %s", err.Error())
+	}
+
+	// Wait for the operation to complete
+	waitErr := appEngineOperationWait(config.clientAppEngine, op, pid, "App Engine app to update")
+	if waitErr != nil {
+		return waitErr
+	}
+	log.Printf("[DEBUG] Updated App Engine App")
+
+	return resourceAppEngineApplicationRead(d, meta)
+}
+
+func resourceAppEngineApplicationDelete(d *schema.ResourceData, meta interface{}) error {
+	log.Println("[DEBUG] App Engine applications cannot be destroyed once created. The project must be deleted to delete the application.")
```

Comment:
it doesn't really matter since from my memory log levels only apply to tf core and not the providers, but if they _did_ make a difference, i would say this should be a WARNING like we do for KmsKeyRing